File: android/src/main/java/com/expogeofencing/ExpoGeofencingModule.kt
package com.expogeofencing

import android.Manifest
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.Looper
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.*
import com.google.android.gms.tasks.Task
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.Enumerable
import android.app.Activity
import android.content.ComponentName
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat

class GeofenceRegionRecord : Record {
  @Field
  val id: String = ""
  
  @Field
  val latitude: Double = 0.0
  
  @Field
  val longitude: Double = 0.0
  
  @Field
  val radius: Float = 0f
  
  @Field
  val notifyOnEntry: Boolean = true
  
  @Field
  val notifyOnExit: Boolean = true
}

class LocationUpdateOptionsRecord : Record {
  @Field
  val interval: Long = 10000
  
  @Field
  val fastestInterval: Long = 5000
  
  @Field
  val smallestDisplacement: Float = 0f
  
  @Field
  val priority: String = "high_accuracy"
}

enum class ActivityType(val value: String) : Enumerable {
  STILL("still"),
  WALKING("walking"),
  RUNNING("running"),
  DRIVING("driving"),
  CYCLING("cycling"),
  UNKNOWN("unknown")
}

class ExpoGeofencingModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private lateinit var fusedLocationClient: FusedLocationProviderClient
  private lateinit var geofencingClient: GeofencingClient
  private lateinit var activityRecognitionClient: ActivityRecognitionClient
  private lateinit var batteryOptimizationHelper: BatteryOptimizationHelper
  
  private var geofenceList = mutableListOf<Geofence>()
  private var locationCallback: LocationCallback? = null
  private var activityCallback: PendingIntent? = null
  private var isMonitoringActive = false
  private var healthCheckInterval = 5 * 60 * 1000L // 5 minutes

  override fun definition() = ModuleDefinition {
    Name("ExpoGeofencing")

    Events("onGeofenceEvent", "onActivityRecognition", "onLocationUpdate", "onHealthAlert", "onServiceStatus")

    OnCreate {
      fusedLocationClient = LocationServices.getFusedLocationProviderClient(context)
      geofencingClient = LocationServices.getGeofencingClient(context)
      activityRecognitionClient = ActivityRecognition.getClient(context)
      batteryOptimizationHelper = BatteryOptimizationHelper(context)
    }

    AsyncFunction("addGeofence") { region: GeofenceRegionRecord, promise: Promise ->
      if (!validateGeofenceRegion(region)) {
        promise.reject("INVALID_REGION", "Invalid geofence region parameters", null)
        return@AsyncFunction
      }
      if (!hasLocationPermission()) {
        promise.reject("PERMISSION_DENIED", "Location permission not granted", null)
        return@AsyncFunction
      }

      val geofence = Geofence.Builder()
        .setRequestId(region.id)
        .setCircularRegion(region.latitude, region.longitude, region.radius)
        .setExpirationDuration(Geofence.NEVER_EXPIRE)
        .setTransitionTypes(
          (if (region.notifyOnEntry) Geofence.GEOFENCE_TRANSITION_ENTER else 0) or
          (if (region.notifyOnExit) Geofence.GEOFENCE_TRANSITION_EXIT else 0)
        )
        .build()

      geofenceList.add(geofence)

      val geofencingRequest = GeofencingRequest.Builder()
        .setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER)
        .addGeofence(geofence)
        .build()

      val pendingIntent = createGeofencePendingIntent()

      geofencingClient.addGeofences(geofencingRequest, pendingIntent)
        .addOnSuccessListener { 
          startPersistenceServiceIfNeeded()
          sendServiceStatusUpdate()
          promise.resolve(null)
        }
        .addOnFailureListener { exception ->
          promise.reject("GEOFENCE_ERROR", "Failed to add geofence: ${exception.message}", exception)
        }
    }

    AsyncFunction("removeGeofence") { regionId: String, promise: Promise ->
      geofencingClient.removeGeofences(listOf(regionId))
        .addOnSuccessListener {
          geofenceList.removeAll { it.requestId == regionId }
          sendServiceStatusUpdate()
          promise.resolve(null)
        }
        .addOnFailureListener { exception ->
          promise.reject("GEOFENCE_ERROR", "Failed to remove geofence: ${exception.message}", exception)
        }
    }

    AsyncFunction("removeAllGeofences") { promise: Promise ->
      geofencingClient.removeGeofences(createGeofencePendingIntent())
        .addOnSuccessListener {
          geofenceList.clear()
          sendServiceStatusUpdate()
          promise.resolve(null)
        }
        .addOnFailureListener { exception ->
          promise.reject("GEOFENCE_ERROR", "Failed to remove all geofences: ${exception.message}", exception)
        }
    }

    AsyncFunction("startActivityRecognition") { intervalMs: Long, promise: Promise ->
      if (intervalMs < 1000) {
        promise.reject("INVALID_INTERVAL", "Activity recognition interval must be at least 1000ms", null)
        return@AsyncFunction
      }
      val intent = Intent(context, ActivityRecognitionReceiver::class.java)
      activityCallback = PendingIntent.getBroadcast(
        context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
      )

      activityRecognitionClient.requestActivityUpdates(intervalMs, activityCallback!!)
        .addOnSuccessListener {
          promise.resolve(null)
        }
        .addOnFailureListener { exception ->
          promise.reject("ACTIVITY_ERROR", "Failed to start activity recognition: ${exception.message}", exception)
        }
    }

    AsyncFunction("stopActivityRecognition") { promise: Promise ->
      activityCallback?.let { callback ->
        activityRecognitionClient.removeActivityUpdates(callback)
          .addOnSuccessListener {
            promise.resolve(null)
          }
          .addOnFailureListener { exception ->
            promise.reject("ACTIVITY_ERROR", "Failed to stop activity recognition: ${exception.message}", exception)
          }
      } ?: promise.resolve(null)
    }

    AsyncFunction("startLocationUpdates") { options: LocationUpdateOptionsRecord, promise: Promise ->
      if (!validateLocationUpdateOptions(options)) {
        promise.reject("INVALID_OPTIONS", "Invalid location update options", null)
        return@AsyncFunction
      }
      if (!hasLocationPermission()) {
        promise.reject("PERMISSION_DENIED", "Location permission not granted", null)
        return@AsyncFunction
      }

      val locationRequest = LocationRequest.Builder(
        getPriorityFromString(options.priority),
        options.interval
      )
        .setMinUpdateIntervalMillis(options.fastestInterval)
        .setMinUpdateDistanceMeters(options.smallestDisplacement)
        .build()

      locationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
          locationResult.locations.forEach { location ->
            sendEvent("onLocationUpdate", mapOf(
              "latitude" to location.latitude,
              "longitude" to location.longitude,
              "accuracy" to location.accuracy,
              "timestamp" to location.time
            ))
          }
        }
      }

      fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback!!, Looper.getMainLooper())
        .addOnSuccessListener {
          isMonitoringActive = true
          startPersistenceServiceIfNeeded()
          promise.resolve(null)
        }
        .addOnFailureListener { exception ->
          promise.reject("LOCATION_ERROR", "Failed to start location updates: ${exception.message}", exception)
        }
    }

    AsyncFunction("stopLocationUpdates") { promise: Promise ->
      locationCallback?.let { callback ->
        fusedLocationClient.removeLocationUpdates(callback)
        locationCallback = null
        isMonitoringActive = false
        promise.resolve(null)
      } ?: promise.resolve(null)
    }

    // New functions for enhanced features
    AsyncFunction("checkBatteryOptimization") { promise: Promise ->
      val isOptimized = !batteryOptimizationHelper.isBatteryOptimizationDisabled()
      val deviceInstructions = batteryOptimizationHelper.getDeviceSpecificBatteryOptimizationInstructions()
      val locationEnabled = batteryOptimizationHelper.checkLocationServicesEnabled()
      
      promise.resolve(mapOf(
        "isBatteryOptimized" to isOptimized,
        "isLocationEnabled" to locationEnabled,
        "deviceInstructions" to deviceInstructions
      ))
    }

    AsyncFunction("requestBatteryOptimizationDisable") { promise: Promise ->
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        try {
          val activity = appContext.currentActivity as? Activity
          if (activity != null) {
            batteryOptimizationHelper.requestBatteryOptimizationDisable(activity, promise)
          } else {
            promise.reject("NO_ACTIVITY", "No current activity available", null)
          }
        } catch (e: Exception) {
          promise.reject("BATTERY_OPTIMIZATION_ERROR", "Failed to request battery optimization disable: ${e.message}", e)
        }
      } else {
        promise.resolve(true) // Not needed on older versions
      }
    }

    AsyncFunction("openLocationSettings") { promise: Promise ->
      try {
        batteryOptimizationHelper.openLocationSettings()
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("SETTINGS_ERROR", "Failed to open location settings: ${e.message}", e)
      }
    }

    AsyncFunction("setHealthCheckInterval") { intervalMs: Long, promise: Promise ->
      if (intervalMs < 60000) { // Minimum 1 minute
        promise.reject("INVALID_INTERVAL", "Health check interval must be at least 60000ms (1 minute)", null)
        return@AsyncFunction
      }
      
      healthCheckInterval = intervalMs
      
      // Restart health monitoring with new interval
      if (isMonitoringActive) {
        LocationHealthMonitor.stopMonitoring(context)
        LocationHealthMonitor.startMonitoring(context, intervalMs)
      }
      
      promise.resolve(null)
    }

    AsyncFunction("getServiceStatus") { promise: Promise ->
      val status = mapOf(
        "isMonitoringActive" to isMonitoringActive,
        "activeGeofences" to geofenceList.size,
        "isBatteryOptimized" to !batteryOptimizationHelper.isBatteryOptimizationDisabled(),
        "isLocationEnabled" to batteryOptimizationHelper.checkLocationServicesEnabled(),
        "healthCheckInterval" to healthCheckInterval,
        "hasLocationPermission" to hasLocationPermission(),
        "hasBackgroundLocationPermission" to hasBackgroundLocationPermission(),
        "hasActivityRecognitionPermission" to hasActivityRecognitionPermission()
      )
      
      promise.resolve(status)
    }

    AsyncFunction("forceLocationCheck") { promise: Promise ->
      if (!hasLocationPermission()) {
        promise.reject("PERMISSION_DENIED", "Location permission not granted", null)
        return@AsyncFunction
      }

      val locationRequest = LocationRequest.Builder(
        Priority.PRIORITY_HIGH_ACCURACY,
        1000
      ).setMaxUpdateDelayMillis(5000).build()

      fusedLocationClient.requestLocationUpdates(
        locationRequest,
        object : LocationCallback() {
          override fun onLocationResult(locationResult: LocationResult) {
            fusedLocationClient.removeLocationUpdates(this)
            val location = locationResult.lastLocation
            if (location != null) {
              promise.resolve(mapOf(
                "latitude" to location.latitude,
                "longitude" to location.longitude,
                "accuracy" to location.accuracy,
                "timestamp" to location.time
              ))
            } else {
              promise.reject("NO_LOCATION", "Unable to get current location", null)
            }
          }
        },
        Looper.getMainLooper()
      )
    }
  }

  private fun hasLocationPermission(): Boolean {
    return ActivityCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_FINE_LOCATION
    ) == PackageManager.PERMISSION_GRANTED
  }

  private fun createGeofencePendingIntent(): PendingIntent {
    val intent = Intent(context, GeofenceBroadcastReceiver::class.java)
    return PendingIntent.getBroadcast(
      context,
      0,
      intent,
      PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    )
  }

  private fun getPriorityFromString(priority: String): Int {
    return when (priority) {
      "high_accuracy" -> Priority.PRIORITY_HIGH_ACCURACY
      "balanced" -> Priority.PRIORITY_BALANCED_POWER_ACCURACY
      "low_power" -> Priority.PRIORITY_LOW_POWER
      "no_power" -> Priority.PRIORITY_PASSIVE
      else -> Priority.PRIORITY_HIGH_ACCURACY
    }
  }

  private fun validateGeofenceRegion(region: GeofenceRegionRecord): Boolean {
    return region.id.isNotEmpty() &&
           region.latitude >= -90.0 && region.latitude <= 90.0 &&
           region.longitude >= -180.0 && region.longitude <= 180.0 &&
           region.radius > 0 && region.radius <= 10000 // Max 10km radius
  }

  private fun validateLocationUpdateOptions(options: LocationUpdateOptionsRecord): Boolean {
    return options.interval >= 1000 && // Min 1 second
           options.fastestInterval >= 500 && // Min 500ms
           options.smallestDisplacement >= 0 &&
           options.priority in listOf("high_accuracy", "balanced", "low_power", "no_power")
  }

  private fun hasBackgroundLocationPermission(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      ContextCompat.checkSelfPermission(
        context,
        android.Manifest.permission.ACCESS_BACKGROUND_LOCATION
      ) == PackageManager.PERMISSION_GRANTED
    } else {
      hasLocationPermission() // Background permission included in location permission on older versions
    }
  }

  private fun hasActivityRecognitionPermission(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      ContextCompat.checkSelfPermission(
        context,
        android.Manifest.permission.ACTIVITY_RECOGNITION
      ) == PackageManager.PERMISSION_GRANTED
    } else {
      true // Not required on older versions
    }
  }

  private fun startPersistenceServiceIfNeeded() {
    if (geofenceList.isNotEmpty() || isMonitoringActive) {
      GeofencingPersistenceService.startService(context)
    }
  }

  private fun sendServiceStatusUpdate() {
    val status = mapOf(
      "isMonitoringActive" to isMonitoringActive,
      "activeGeofences" to geofenceList.size,
      "timestamp" to System.currentTimeMillis()
    )
    sendEvent("onServiceStatus", status)
  }
}

File: ios/ExpoGeofencing/ExpoGeofencingModule.swift
import ExpoModulesCore
import CoreLocation
import CoreMotion
import UserNotifications
import UIKit
import BackgroundTasks

public class ExpoGeofencingModule: Module {
  private var locationManager: CLLocationManager?
  private var motionActivityManager: CMMotionActivityManager?
  private var geofenceRegions: [String: CLCircularRegion] = [:]
  private var isMonitoringActivity = false
  private var isMonitoringLocation = false
  private var healthCheckTimer: Timer?
  private var healthCheckInterval: TimeInterval = 5 * 60 // 5 minutes
  private var lastLocationTime: Date = Date()
  private var lastAccurateLocationTime: Date = Date()
  private var backgroundTask: UIBackgroundTaskIdentifier = .invalid
  
  public func definition() -> ModuleDefinition {
    Name("ExpoGeofencing")
    
    Events("onGeofenceEvent", "onActivityRecognition", "onLocationUpdate", "onHealthAlert", "onServiceStatus")
    
    OnCreate {
      locationManager = CLLocationManager()
      locationManager?.delegate = self
      motionActivityManager = CMMotionActivityManager()
      setupNotifications()
      setupBackgroundTasks()
    }
    
    AsyncFunction("addGeofence") { (region: GeofenceRegion, promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      guard self.validateGeofenceRegion(region) else {
        promise.reject("INVALID_REGION", "Invalid geofence region parameters")
        return
      }
      
      guard locationManager.authorizationStatus == .authorizedAlways ||
            locationManager.authorizationStatus == .authorizedWhenInUse else {
        promise.reject("PERMISSION_DENIED", "Location permission not granted")
        return
      }
      
      let center = CLLocationCoordinate2D(latitude: region.latitude, longitude: region.longitude)
      let geofenceRegion = CLCircularRegion(center: center, radius: region.radius, identifier: region.id)
      
      geofenceRegion.notifyOnEntry = region.notifyOnEntry
      geofenceRegion.notifyOnExit = region.notifyOnExit
      
      geofenceRegions[region.id] = geofenceRegion
      locationManager.startMonitoring(for: geofenceRegion)
      
      self.startHealthMonitoringIfNeeded()
      self.sendServiceStatusUpdate()
      promise.resolve(nil)
    }
    
    AsyncFunction("removeGeofence") { (regionId: String, promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      if let region = geofenceRegions[regionId] {
        locationManager.stopMonitoring(for: region)
        geofenceRegions.removeValue(forKey: regionId)
      }
      
      promise.resolve(nil)
    }
    
    AsyncFunction("removeAllGeofences") { (promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      for region in geofenceRegions.values {
        locationManager.stopMonitoring(for: region)
      }
      geofenceRegions.removeAll()
      
      promise.resolve(nil)
    }
    
    AsyncFunction("startActivityRecognition") { (intervalMs: Int, promise: Promise) in
      guard let motionActivityManager = motionActivityManager else {
        promise.reject("MOTION_MANAGER_ERROR", "Motion activity manager not initialized")
        return
      }
      
      guard CMMotionActivityManager.isActivityAvailable() else {
        promise.reject("ACTIVITY_NOT_AVAILABLE", "Activity recognition not available on this device")
        return
      }
      
      let interval = TimeInterval(intervalMs / 1000)
      isMonitoringActivity = true
      
      motionActivityManager.startActivityUpdates(to: OperationQueue.main) { [weak self] activity in
        guard let self = self, let activity = activity, self.isMonitoringActivity else { return }
        
        let activityType = self.getActivityType(from: activity)
        let confidence = self.getConfidence(from: activity)
        
        let eventData: [String: Any] = [
          "activity": activityType,
          "confidence": confidence,
          "timestamp": Int(activity.timestamp.timeIntervalSince1970 * 1000)
        ]
        
        self.sendEvent("onActivityRecognition", eventData)
      }
      
      promise.resolve(nil)
    }
    
    AsyncFunction("stopActivityRecognition") { (promise: Promise) in
      guard let motionActivityManager = motionActivityManager else {
        promise.reject("MOTION_MANAGER_ERROR", "Motion activity manager not initialized")
        return
      }
      
      isMonitoringActivity = false
      motionActivityManager.stopActivityUpdates()
      
      promise.resolve(nil)
    }
    
    AsyncFunction("startLocationUpdates") { (options: LocationUpdateOptions, promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      guard self.validateLocationUpdateOptions(options) else {
        promise.reject("INVALID_OPTIONS", "Invalid location update options")
        return
      }
      
      guard locationManager.authorizationStatus == .authorizedAlways ||
            locationManager.authorizationStatus == .authorizedWhenInUse else {
        promise.reject("PERMISSION_DENIED", "Location permission not granted")
        return
      }
      
      locationManager.desiredAccuracy = getAccuracy(from: options.priority)
      locationManager.distanceFilter = CLLocationDistance(options.smallestDisplacement)
      
      locationManager.startUpdatingLocation()
      isMonitoringLocation = true
      
      self.startHealthMonitoringIfNeeded()
      self.sendServiceStatusUpdate()
      promise.resolve(nil)
    }
    
    AsyncFunction("stopLocationUpdates") { (promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      locationManager.stopUpdatingLocation()
      isMonitoringLocation = false
      
      self.sendServiceStatusUpdate()
      promise.resolve(nil)
    }
    
    // Enhanced iOS-specific functions
    AsyncFunction("checkSystemStatus") { (promise: Promise) in
      let authStatus = locationManager?.authorizationStatus ?? .notDetermined
      let backgroundRefreshStatus = UIApplication.shared.backgroundRefreshStatus
      let locationServicesEnabled = CLLocationManager.locationServicesEnabled()
      
      let status: [String: Any] = [
        "locationPermission": self.getLocationPermissionStatus(authStatus),
        "backgroundRefreshStatus": self.getBackgroundRefreshStatus(backgroundRefreshStatus),
        "locationServicesEnabled": locationServicesEnabled,
        "isMonitoringActive": isMonitoringLocation || !geofenceRegions.isEmpty,
        "activeGeofences": geofenceRegions.count,
        "activityRecognitionAvailable": CMMotionActivityManager.isActivityAvailable(),
        "healthCheckInterval": healthCheckInterval
      ]
      
      promise.resolve(status)
    }
    
    AsyncFunction("requestPermissions") { (promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      // Request location permissions
      locationManager.requestAlwaysAuthorization()
      
      // Request notification permissions
      UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
        if let error = error {
          promise.reject("NOTIFICATION_PERMISSION_ERROR", "Failed to request notification permission: \(error.localizedDescription)")
          return
        }
        promise.resolve(granted)
      }
    }
    
    AsyncFunction("setHealthCheckInterval") { (intervalMs: Int, promise: Promise) in
      guard intervalMs >= 60000 else { // Minimum 1 minute
        promise.reject("INVALID_INTERVAL", "Health check interval must be at least 60000ms (1 minute)")
        return
      }
      
      healthCheckInterval = TimeInterval(intervalMs / 1000)
      
      // Restart health monitoring with new interval
      self.stopHealthMonitoring()
      self.startHealthMonitoringIfNeeded()
      
      promise.resolve(nil)
    }
    
    AsyncFunction("forceLocationCheck") { (promise: Promise) in
      guard let locationManager = locationManager else {
        promise.reject("LOCATION_MANAGER_ERROR", "Location manager not initialized")
        return
      }
      
      guard locationManager.authorizationStatus == .authorizedAlways ||
            locationManager.authorizationStatus == .authorizedWhenInUse else {
        promise.reject("PERMISSION_DENIED", "Location permission not granted")
        return
      }
      
      locationManager.requestLocation()
      promise.resolve(nil)
    }
  }
  
  private func getActivityType(from activity: CMMotionActivity) -> String {
    if activity.stationary {
      return "still"
    } else if activity.walking {
      return "walking"
    } else if activity.running {
      return "running"
    } else if activity.automotive {
      return "driving"
    } else if activity.cycling {
      return "cycling"
    } else {
      return "unknown"
    }
  }
  
  private func getConfidence(from activity: CMMotionActivity) -> Int {
    switch activity.confidence {
    case .low:
      return 25
    case .medium:
      return 50
    case .high:
      return 75
    @unknown default:
      return 0
    }
  }
  
  private func getAccuracy(from priority: String) -> CLLocationAccuracy {
    switch priority {
    case "high_accuracy":
      return kCLLocationAccuracyBest
    case "balanced":
      return kCLLocationAccuracyNearestTenMeters
    case "low_power":
      return kCLLocationAccuracyHundredMeters
    case "no_power":
      return kCLLocationAccuracyThreeKilometers
    default:
      return kCLLocationAccuracyBest
    }
  }
  
  // MARK: - Validation Functions
  private func validateGeofenceRegion(_ region: GeofenceRegion) -> Bool {
    return !region.id.isEmpty &&
           region.latitude >= -90.0 && region.latitude <= 90.0 &&
           region.longitude >= -180.0 && region.longitude <= 180.0 &&
           region.radius > 0 && region.radius <= 10000 // Max 10km radius
  }
  
  private func validateLocationUpdateOptions(_ options: LocationUpdateOptions) -> Bool {
    return options.interval >= 1000 && // Min 1 second
           options.fastestInterval >= 500 && // Min 500ms
           options.smallestDisplacement >= 0 &&
           ["high_accuracy", "balanced", "low_power", "no_power"].contains(options.priority)
  }
  
  // MARK: - System Status Functions
  private func getLocationPermissionStatus(_ status: CLAuthorizationStatus) -> String {
    switch status {
    case .notDetermined:
      return "not_determined"
    case .restricted:
      return "restricted"
    case .denied:
      return "denied"
    case .authorizedWhenInUse:
      return "when_in_use"
    case .authorizedAlways:
      return "always"
    @unknown default:
      return "unknown"
    }
  }
  
  private func getBackgroundRefreshStatus(_ status: UIBackgroundRefreshStatus) -> String {
    switch status {
    case .restricted:
      return "restricted"
    case .denied:
      return "denied"
    case .available:
      return "available"
    @unknown default:
      return "unknown"
    }
  }
  
  // MARK: - Health Monitoring Functions
  private func startHealthMonitoringIfNeeded() {
    guard (isMonitoringLocation || !geofenceRegions.isEmpty) && healthCheckTimer == nil else {
      return
    }
    
    healthCheckTimer = Timer.scheduledTimer(withTimeInterval: healthCheckInterval, repeats: true) { _ in
      self.performHealthCheck()
    }
    
    // Also perform an initial check
    performHealthCheck()
  }
  
  private func stopHealthMonitoring() {
    healthCheckTimer?.invalidate()
    healthCheckTimer = nil
  }
  
  private func performHealthCheck() {
    let currentTime = Date()
    let timeSinceLastLocation = currentTime.timeIntervalSince(lastLocationTime)
    let timeSinceAccurateLocation = currentTime.timeIntervalSince(lastAccurateLocationTime)
    
    // Check if location services are still enabled
    if !CLLocationManager.locationServicesEnabled() {
      sendHealthAlert(title: "Location Services Disabled", message: "Please enable location services to continue geofencing")
      return
    }
    
    // Check authorization status
    if let locationManager = locationManager {
      if locationManager.authorizationStatus == .denied || locationManager.authorizationStatus == .restricted {
        sendHealthAlert(title: "Location Permission Denied", message: "Location permission is required for geofencing")
        return
      }
    }
    
    // Check background refresh status
    if UIApplication.shared.backgroundRefreshStatus == .denied {
      sendHealthAlert(title: "Background Refresh Disabled", message: "Enable background refresh for reliable geofencing")
    }
    
    // Check location timeouts
    if timeSinceLastLocation > 10 * 60 { // 10 minutes
      sendHealthAlert(title: "Location Timeout", message: "No location updates received recently")
      // Try to get current location
      locationManager?.requestLocation()
    }
    
    if timeSinceAccurateLocation > 30 * 60 { // 30 minutes
      sendHealthAlert(title: "GPS Accuracy Low", message: "Recent location updates are not accurate enough")
    }
    
    // Send periodic health check location request
    if isMonitoringLocation {
      locationManager?.requestLocation()
    }
  }
  
  private func sendHealthAlert(title: String, message: String) {
    let alertData: [String: Any] = [
      "type": "health_alert",
      "title": title,
      "message": message,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000)
    ]
    
    sendEvent("onHealthAlert", alertData)
    
    // Also send local notification if app is in background
    if UIApplication.shared.applicationState != .active {
      sendLocalNotification(title: title, message: message)
    }
  }
  
  private func sendServiceStatusUpdate() {
    let status: [String: Any] = [
      "isMonitoringActive": isMonitoringLocation || !geofenceRegions.isEmpty,
      "activeGeofences": geofenceRegions.count,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000)
    ]
    sendEvent("onServiceStatus", status)
  }
  
  // MARK: - Notification Functions
  private func setupNotifications() {
    UNUserNotificationCenter.current().delegate = self
  }
  
  private func sendLocalNotification(title: String, message: String) {
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = message
    content.sound = .default
    
    let request = UNNotificationRequest(
      identifier: "geofencing_health_\(UUID().uuidString)",
      content: content,
      trigger: nil
    )
    
    UNUserNotificationCenter.current().add(request)
  }
  
  // MARK: - Background Task Functions
  private func setupBackgroundTasks() {
    if #available(iOS 13.0, *) {
      BGTaskScheduler.shared.register(forTaskWithIdentifier: "com.expogeofencing.healthcheck", using: nil) { task in
        self.handleBackgroundHealthCheck(task: task as! BGAppRefreshTask)
      }
    }
  }
  
  @available(iOS 13.0, *)
  private func handleBackgroundHealthCheck(task: BGAppRefreshTask) {
    task.expirationHandler = {
      task.setTaskCompleted(success: false)
    }
    
    performHealthCheck()
    
    // Schedule next background task
    scheduleBackgroundHealthCheck()
    
    task.setTaskCompleted(success: true)
  }
  
  @available(iOS 13.0, *)
  private func scheduleBackgroundHealthCheck() {
    let request = BGAppRefreshTaskRequest(identifier: "com.expogeofencing.healthcheck")
    request.earliestBeginDate = Date(timeIntervalSinceNow: healthCheckInterval)
    
    try? BGTaskScheduler.shared.submit(request)
  }
}

extension ExpoGeofencingModule: CLLocationManagerDelegate {
  public func locationManager(_ manager: CLLocationManager, didEnterRegion region: CLRegion) {
    handleGeofenceEvent(region: region, eventType: "enter")
  }
  
  public func locationManager(_ manager: CLLocationManager, didExitRegion region: CLRegion) {
    handleGeofenceEvent(region: region, eventType: "exit")
  }
  
  public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    guard let location = locations.last else { return }
    
    lastLocationTime = Date()
    
    if location.horizontalAccuracy <= 50 { // Consider accurate if within 50 meters
      lastAccurateLocationTime = Date()
    }
    
    let eventData: [String: Any] = [
      "latitude": location.coordinate.latitude,
      "longitude": location.coordinate.longitude,
      "accuracy": location.horizontalAccuracy,
      "timestamp": Int(location.timestamp.timeIntervalSince1970 * 1000),
      "isHealthCheck": true
    ]
    
    sendEvent("onLocationUpdate", eventData)
  }
  
  public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    let alertData: [String: Any] = [
      "type": "location_error",
      "title": "Location Error",
      "message": "Failed to get location: \(error.localizedDescription)",
      "timestamp": Int(Date().timeIntervalSince1970 * 1000)
    ]
    
    sendEvent("onHealthAlert", alertData)
  }
  
  private func handleGeofenceEvent(region: CLRegion, eventType: String) {
    guard let location = locationManager?.location else { return }
    
    let eventData: [String: Any] = [
      "regionId": region.identifier,
      "eventType": eventType,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000),
      "latitude": location.coordinate.latitude,
      "longitude": location.coordinate.longitude
    ]
    
    sendEvent("onGeofenceEvent", eventData)
  }
}

struct GeofenceRegion: Record {
  @Field var id: String = ""
  @Field var latitude: Double = 0.0
  @Field var longitude: Double = 0.0
  @Field var radius: Double = 0.0
  @Field var notifyOnEntry: Bool = true
  @Field var notifyOnExit: Bool = true
}

struct LocationUpdateOptions: Record {
  @Field var interval: Int = 10000
  @Field var fastestInterval: Int = 5000
  @Field var smallestDisplacement: Double = 0.0
  @Field var priority: String = "high_accuracy"
}

// MARK: - UNUserNotificationCenterDelegate
extension ExpoGeofencingModule: UNUserNotificationCenterDelegate {
  public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
    completionHandler([.banner, .sound, .badge])
  }
}